### PR TITLE
worker: Add common initialize methods for external services

### PR DESCRIPTION
**Pull request by efritz**
This PR:

- adds external service initialization methods for:
  - the frontend database
  - the codeintel-db
  - common code intelligence stores and clients (dbstore, lsifstore, gitserver client)
- sets up authz permissions

Nothing currently uses these, but there was some non-trivial memoization work I wanted to do separately before moving any actual behavior into this service.

### CODE REVIEW

File: enterprise/cmd/worker/internal/codeintel/gitserver.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/sourcegraph/sourcegraph/cmd/worker/shared"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/gitserver"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+	"github.com/sourcegraph/sourcegraph/internal/trace"
+)
+
+// InitGitserverClient initializes and returns a gitserver client.
+func InitGitserverClient() (*sql.DB, error) {
```
Comment (efritz) [inline enterprise/cmd/worker/internal/codeintel/gitserver.go:17]:
```suggestion
func InitGitserverClient() (*gitserver.Client, error) {
```

File: enterprise/cmd/worker/internal/codeintel/gitserver.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	conn, err := initGitserverClient.Init()
+	return conn.(*sql.DB), err
```
Comment (efritz) [inline enterprise/cmd/worker/internal/codeintel/gitserver.go:19]:
```suggestion
	return conn.(*gitserver.Client), err
```

File: enterprise/cmd/worker/internal/codeintel/dbstore.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/sourcegraph/sourcegraph/cmd/worker/shared"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+	"github.com/sourcegraph/sourcegraph/internal/trace"
+)
+
+// InitDBStore initializes and returns a db store.
```
Comment (efritz) [inline enterprise/cmd/worker/internal/codeintel/dbstore.go:14]:
```suggestion
// InitDBStore initializes and returns a db store instance.
```

File: enterprise/cmd/worker/main.go [hunk 4]
```diff
@@ -14,5 +20,26 @@ func main() {
 		log.Println("enterprise edition")
 	}
 
-	shared.Start(nil)
+	go setAuthProviders()
```
Comment (efritz) [inline enterprise/cmd/worker/main.go:23]: @unknwon When connecting a service to a database, what are the auth setup steps we need to keep in mind? What's in this file today is basically what was added to the precise-code-intel-worker.

Given that everything in this service is running as the "system" and not (directly) on behalf of a user-request, do we need to keep the same setup as the frontend?

Comment (unknwon) [reply enterprise/cmd/worker/main.go:23]: >When connecting a service to a database, what are the auth setup steps we need to keep in mind? What's in this file today is basically what was added to the precise-code-intel-worker.

What you've done inside `setAuthProviders` is what required to do.

>Given that everything in this service is running as the "system" and not (directly) on behalf of a user-request, do we need to keep the same setup as the frontend?

If you're 100% sure at all time, this is all you need to to:

```go
authz.SetProviders(true, []authz.Provider{})
```

Which is what gitserver does: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c64d770/-/blob/cmd/gitserver/main.go#L228

Comment (efritz) [reply enterprise/cmd/worker/main.go:23]: Can you give a quick overview of what the auth providers _do_ under the hood? I assume it's only affecting what repos are visible to a user (requiring an explicit call to construct additional WHERE conditions) (and I assume that user data is stored within a context related to a users request).

Comment (unknwon) [reply enterprise/cmd/worker/main.go:23]: >Can you give a quick overview of what the auth providers do under the hood?

Hmm, speaking of this, I realized here we are talking about authz providers, not auth providers. The former is for authorization (e.g. repo perms), the latter is for authentication (user sign0in).

On the write path, authz providers are used to fetch repo permissions from code hosts (which is only matters currently in repo-updater). On the read path, currently only the _number of authz providers_ is used for checking [conflict with explicitly permissions](https://github.com/sourcegraph/sourcegraph/blob/8d3ebdc631122bbdd47adb0dd77f0c2c52184c13/internal/database/repos_perm.go#L27-L29) and [authz bypass](https://github.com/sourcegraph/sourcegraph/blob/8d3ebdc631122bbdd47adb0dd77f0c2c52184c13/internal/database/repos_perm.go#L36-L40) (the thing you're looking for, and what my above code snippet does).

>I assume it's only affecting what repos are visible to a user (requiring an explicit call to construct additional WHERE conditions) (and I assume that user data is stored within a context related to a users request).

Currently yes and yes.

Comment (efritz) [reply enterprise/cmd/worker/main.go:23]: Thank you for walking me through it!

File: enterprise/cmd/worker/internal/codeintel/dbstore.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+func InitDBStore() (*dbstore.Store, error) {
+	conn, err := initDBStore.Init()
+	return conn.(*dbstore.Store), err
+}
+
+var initDBStore = shared.NewMemoizedConstructor(func() (interface{}, error) {
+	observationContext := &observation.Context{
+		Logger:     log15.Root(),
+		Tracer:     &trace.Tracer{Tracer: opentracing.GlobalTracer()},
+		Registerer: prometheus.DefaultRegisterer,
```
Comment (tsenart) [inline enterprise/cmd/worker/internal/codeintel/dbstore.go:24]: We should add the profiler at some point :) This is running in repo-updater, gitserver, frontend and zoekt.

Comment (tsenart) [reply enterprise/cmd/worker/internal/codeintel/dbstore.go:24]: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/profiler/profiler.go?subtree=true

File: enterprise/cmd/worker/main.go [hunk 6]
```diff
@@ -14,5 +20,28 @@ func main() {
...
+
+	shared.Start(map[string]shared.Task{
```
Comment (tsenart) [inline enterprise/cmd/worker/main.go:25]: nittiry nitty nit: I like the name Job better than Task.

**Review by tsenart — APPROVED** [review]

File: enterprise/cmd/worker/main.go (hunk 6 above)
Comment (efritz) [reply enterprise/cmd/worker/main.go:25]: Will rename in a follow-up PR to keep this one (relatively, minus the typo fix) clean.

Comment (efritz) [reply enterprise/cmd/worker/main.go:25]: https://github.com/sourcegraph/sourcegraph/pull/21837

File: enterprise/cmd/worker/main.go [hunk 7]
```diff
@@ -14,5 +20,28 @@ func main() {
...
+		// Empty for now
+	})
+}
+
+// setAuthProviders waits for the database to be initialized, then periodically refreshes the
+// global authz providers. This changes the repositories that are visible for reads based on the
+// current actor stored in an operation's context, which is likely an internal actor for many of
+// the tasks configured in this service. This also enables repository update operations to fetch
+// permissions from code hosts.
+func setAuthProviders() {
```
Comment (unknwon) [inline enterprise/cmd/worker/main.go:35]: To avoid confusion, this function should be actually renamed to `setAuthzProviders`

**Pull request merged**